mujoco_py/mjpid.pyx

```diff
@@ -24,77 +81,191 @@ cdef enum USER_DEFINED_ACTUATOR_PARAMS:
 
-    cdef double effort_limit_low = m.actuator_forcerange[id * 2]
-    cdef double effort_limit_high = m.actuator_forcerange[id * 2 + 1]
+    if Kp_cas != 0:
+        # Run a position PID loop and use the result to set a desired velocity signal
+        pos_output = _pid(parameters=PIDParameters(
+            dt_seconds=dt_in_sec,
+            setpoint=d.ctrl[id],
+            feedback=d.actuator_length[id],
+            Kp=Kp_cas,
+            error_deadband=0.0,
+            integral_max_clamp=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_MAX_CLAMP],
+            integral_time_const=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_TIME_CONSTANT],
+            derivative_gain_smoothing=0.0,
+            derivative_time_const=0.0,
+            previous_errors=PIDErrors(
+                error=0.0,
+                derivative_error=0.0,
+                integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR],
+            )))
 
-    if fabs(error) < error_deadband:
-        error = 0.0
+        # Save errors
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_LAST_ERROR] = pos_output.errors.error
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR] = pos_output.errors.integral_error
+        des_vel = Kp_cas * (pos_output.errors.error + pos_output.errors.integral_error)
+    else:
+        # If P gain on position loop is zero, only use the velocity controller
+        des_vel = d.ctrl[id]
 
-    integral_error = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INTEGRAL_ERROR]
-    integral_error += error * dt_in_sec
-    integral_error = fmax(-integral_max_clamp, fmin(integral_max_clamp, integral_error))
+    # Clamp max angular velocity
+    max_qvel = m.actuator_gainprm[id * NGAIN + IDX_CAS_MAX_VEL]
+    des_vel = fmax(-max_qvel, fmin(max_qvel, des_vel))
 
-    last_error = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_LAST_ERROR]
-    cdef double derivative_error = (error - last_error) / dt_in_sec
+    # Apply Exponential Moving Average smoothing to the velocity setpoint
+    ctrl_ema_vel = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_STORED_EMA_SMOOTH_V]
+    smoothing_factor = m.actuator_gainprm[id * NGAIN + IDX_CAS_EMA_SMOOTH_V]
 
-    derivative_error_last = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_DERIVATIVE_ERROR_LAST]
+    smoothed_vel_setpoint = (smoothing_factor * ctrl_ema_vel) + (1 - smoothing_factor) * des_vel
 
-    derivative_error = (1.0 - derivative_gain_smoothing) * derivative_error_last + \
-        derivative_gain_smoothing * derivative_error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_STORED_EMA_SMOOTH_V] = smoothed_vel_setpoint
 
-    cdef double integral_error_term = 0.0
-    if integral_time_const != 0:
-        integral_error_term = integral_error / integral_time_const
+    vel_output = _pid(parameters=PIDParameters(
+        dt_seconds=dt_in_sec,
+        setpoint=smoothed_vel_setpoint,
```

Comment:
this seems a very complex controller: PI output from the first PID, then exp moving avg, then second PID.  I am a bit concern about ema about the pid output, wouldn't it introduce the behavior of the controller?

Reply:
The biggest issue with the EMA is the delay it introduces on the original position reference signal tracking, other than that, the performance quite good.

Reply:
it is perfectly fine for have EMA on controller inputs (like pos, vels, etc), but here we are applying EMA to the first PID output. is it common in cascaded PID controller?